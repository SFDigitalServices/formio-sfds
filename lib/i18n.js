const dot = require('dotmap')

class UIString {
  static key (component, path) {
    return `${component.key}.${path}`
  }

  constructor (value, component, path, context) {
    if (!value) {
      throw new Error('UIString() requires a value as its first argument')
    } else if (!(component instanceof Object)) {
      throw new Error('UIString() requires a component as its second argument')
    } else if (!component.key) {
      throw new Error('UIString() requires a component object with a "key" property')
    }
    Object.assign(this, {
      value,
      component,
      path,
      _context: context
    })
  }

  get type () {
    return 'string'
  }

  get key () {
    return this.path ? UIString.key(this.component, this.path) : this.value
  }

  get context () {
    return this._context && UIString.key(this.component, this._context)
  }
}

class StringInterpolation extends UIString {
  constructor (value, str) {
    const { component, path, context } = str
    super(value, component, path, context)
    this.string = str
  }

  get type () {
    return 'interpolation'
  }

  get key () {
    return this.value
  }

  get context () {
    return super.key
  }
}

module.exports = {
  getCondition,
  getStrings,
  UIString,
  StringInterpolation
}

function getStrings (form) {
  const fields = [
    fieldGetter('label', (label, component, path) => {
      switch (component.type) {
        case 'content': // we don't show labels for content components
        case 'htmlelement': // we don't show labels for HTML elements
        case 'panel': // panels use the title, not the label
          return []
        default:
          if (label && !component.hideLabel) {
            return [new UIString(label, component, path)]
          }
      }
    }),
    fieldGetter('title', (value, component, path) => {
      if (component.type === 'panel' && value) {
        return [new UIString(value, component, path)]
      }
    }),
    fieldGetter('description'),
    // NB: "html" components have a "content" field
    fieldGetter('html', (value, component, path) => {
      if (component.type === 'content' && value) {
        return [new UIString(value, component, path)]
      }
    }),
    // NB: "content" components have an "html" field
    fieldGetter('content', (value, component, path) => {
      if (component.type === 'html' && value) {
        return [new UIString(value, component, path)]
      }
    }),
    fieldGetter('suffix'),
    fieldGetter('values', (values, component, path) => {
      return Array.isArray(values) && values.map(({ label, value }) => {
        return new UIString(label, component, `values.${value}`)
      })
    }),
    fieldGetter('data', (data, component, path) => {
      const { dataSrc } = component
      const { values } = data
      if (Array.isArray(values) && (!dataSrc || dataSrc === 'values')) {
        return values.map(({ label, value }) => {
          return new UIString(label, component, `values.${value}`)
        })
      } else {
        console.info('Skipping data for component "%s" (dataSrc: "%s"), length: %d', component.key, dataSrc, values.length)
      }
    }),
    fieldGetter('validate', ({ customMessage, custom }, component) => {
      const strings = []
      if (customMessage) {
        strings.push(new UIString(customMessage, component, 'validate.customMessage'))
      }

      if (custom) {
        /**
         * XXX These patterns match anything that _looks_ like
         * a quoted string. We're not doing anything fancy
         * here, and quotes within strings will almost
         * certainly confuse it. For example:
         *
         * `valid = data['x'] || "it's foo"`
         *
         * will yield two "strings" `x` and `] || "it`.
         *
         * If we're only ever going to run this server-side,
         * then we could parse it as JS and only look for
         * string expressions that aren't property accessors.
         */
        const matches = [
          ...custom.matchAll(/."([^"]+)"/g),
          ...custom.matchAll(/.'([^']+)'/g)
        ]
        // eslint-disable-next-line no-unused-vars
        for (const [str, substr] of matches) {
          if (str.startsWith('[') || /^\d+$/.test(substr)) {
            // console.warn('ignoring possible property accessor: "%s"', str)
            continue
          }
          strings.push(new UIString(substr, component, null, 'validate.custom'))
        }
      }
      return strings
    }),
    fieldGetter('customError', (customError, component, path) => {
      return customError
        ? [new UIString(customError, component, path)]
        : []
    }),
    fieldGetter('properties.backTitle', (title, component) => {
      return title
        ? [new UIString(title, component, 'form.backTitle')]
        : []
    })
  ]

  const all = []
  eachComponent(form.components, (component, index, parents) => {
    for (const getFieldStrings of fields) {
      const strings = getFieldStrings(component)
        .filter(str => str && str.value)
      for (const str of strings) {
        Object.assign(str, { parents })
      }
      all.push(...strings)
    }
  })

  for (const str of all) {
    str.value.replace(/\$t\((.*?)\)/g, (_, key) => {
      all.push(new StringInterpolation(key, str))
    })
  }

  return all

  function fieldGetter (path, get) {
    if (get) {
      return component => {
        const value = dot.get(component, path)
        return value ? get(value, component, path) || [] : []
      }
    } else {
      return component => {
        const value = dot.get(component, path)
        return value ? [new UIString(value, component, path)] : []
      }
    }
  }
}

function eachComponent (components, iter, parents = []) {
  for (const [index, component] of Object.entries(components)) {
    const next = iter(component, index, parents || [])
    if (next === true) return component
    else if (next === false) return
    let children = component.components
    if (!children && component.columns) {
      children = component.columns.reduce((list, column) => list.concat(column.components), [])
    }
    if (children && children.length) {
      eachComponent(children, iter, parents.concat(component))
    }
  }
}

function getCondition (component) {
  const { customConditional, conditional } = component
  return customConditional || (conditional && conditional.show ? conditional : null)
}
