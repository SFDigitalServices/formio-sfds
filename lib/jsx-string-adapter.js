import { html } from 'jsx-pragmatic/dist/module/renderers/html'

export { node as h, Fragment } from 'jsx-pragmatic/dist/module/node'

export const renderer = html()

export function template (f) {
  return (...args) => f(...args).render(renderer)
}
