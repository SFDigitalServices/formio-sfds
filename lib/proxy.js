import { URL } from 'url'
import fetch from 'node-fetch'
import { JSDOM } from 'jsdom'
import packageJson from '../package.json'

export const LIVE_URL = 'https://sf.gov'

export const environments = {
  live: LIVE_URL,
  test: 'https://test-sfgov.pantheonsite.io'
}

export class Proxy {
  constructor (options) {
    const { env = 'live', base } = options || {}
    this.base = options.url || base || environments[env] || LIVE_URL
    this.options = options
  }

  fetch (uri) {
    const url = /^https?:/.test(uri)
      ? uri
      : new URL(uri, this.base)
    console.warn('fetching:', url)
    return fetch(url)
      .then(async res => {
        this.response = res
        this.url = url
        this.body = await res.text()
        this.dom = new JSDOM(this.body)
        this.document = this.dom.window.document
        const base = this.document.createElement('base')
        base.setAttribute('href', String(this.base))
        this.document.head.insertBefore(base, this.document.head.firstChild)
        return this
      })
  }

  transform (query) {
    const {
      version,
      formioSFDSVersion = version,
      formiojsVersion,
      lang,
      env = 'live'
    } = query

    const { document, url } = this

    let { source, options } = query

    this.setFormioSFDSVersion(formioSFDSVersion)

    if (formiojsVersion) {
      this.setFormioJSVersion(formiojsVersion)
    }

    const div = document.querySelector('[data-source]')
    if (div) {
      if (source) {
        div.setAttribute('data-source', source)
      } else {
        source = div.getAttribute('data-source')
      }

      if (lang) {
        div.setAttribute('lang', lang)
      }

      if (options) {
        const optstr = options instanceof Object ? JSON.stringify(options) : options
        div.setAttribute('data-options', optstr)
      } else {
        options = div.getAttribute('data-options')
      }

      const info = this.element('div', { class: 'formio-sfds' })
      info.innerHTML = `
        <details class="my-4 notranslate" lang="en" translate="no">
          <summary class="m-0">⚠️ This is a form preview</summary>
          <div class="bg-blue-1 p-2 border-1 border-bright-blue round-bottom-1">
            <p class="mt-0 mb-1">
              <b>You are viewing this form in a simulated sf.gov environment.</b>
              The form should function as expected, and styles should reflect
              what you will see when this version of <code>formio-sfds</code>
              is used on sf.gov. Some additional information is listed below.
            </p>
            <dl class="m-0">
              <dt><b>Fetched URL</b></dt>
              <dd class="mb-1 ml-2">
                <a href="${url}"><code>${url}</code></a>
                ${env ? `(via <code>env=${env}</code>)` : ''}
              </dd>

              <dt><b>Form.io JS versions </b></dt>
              <dd class="mb-0 ml-2">
                <a data-script-src="formiojs">
                  <code>formiojs@${formiojsVersion || this.getScriptVersion('formiojs')}</code>
                </a>
              </dd>
              <dd class="mb-0 ml-2">
                <a data-script-src="formio-sfds">
                  <code>formio-sfds@${formioSFDSVersion || this.getScriptVersion('formio-sfds') || packageJson.version}</code>
                </a>
                ${formioSFDSVersion ? '' : ' (local)'}
              </dd>

              <dt><b>Form data source URL</b></dt>
              <dd class="mb-1 ml-2">${
                source ? `<a href="${source}"><code>${source}</code></a>` : 'not provided'
              }</dd>

              <dt><b>Form options</b></dt>
              <dd class="mb-1 ml-2">${
                options ? `<pre>${JSON.stringify(JSON.parse(options), null, 2)}</pre>` : 'none provided'
              }</dd>
            </ul>
          </div>
        </details>
      `.trim()

      div.parentNode.insertBefore(info, div)

      this.appendScript(`
        Array.from(document.querySelectorAll('[data-script-src]')).forEach(link => {
          var pattern = link.getAttribute('data-script-src')
          var script = document.querySelector('script[src*=' + pattern + ']')
          if (script) {
            link.href = script.src
          }
        })
      `)
    } else {
      console.warn('no [data-source] form element found!')
    }
  }

  appendScript (code) {
    const script = this.document.createElement('script')
    script.textContent = `(function () { ${code} })()`
    this.document.body.appendChild(script)
  }

  getScript (srcPattern) {
    const selector = `script[src*="${srcPattern}"]`
    return this.document.querySelector(selector)
  }

  getScriptVersion (srcPattern) {
    const script = this.getScript(srcPattern)
    if (script) {
      const match = script.src.match(/@([^/]+)/)
      return match ? match[1] : undefined
    }
    return undefined
  }

  setFormioSFDSVersion (version) {
    const script = this.getScript('formio-sfds@')
    if (script) {
      if (version) {
        script.setAttribute('src', script.src.replace(/formio-sfds@[^/]+/, `formio-sfds@${version}`))
      } else {
        script.removeAttribute('src')
        script.textContent = `
          document.addEventListener('DOMContentLoaded', () => {
            var script = document.createElement('script')
            var baseUrl = [
              location.protocol, '//', location.hostname,
              location.port ? ':' + location.port : ''
            ].join('')
            var formioSfdsUrl = document.getElementById('formio-sfds-url')
            script.src = baseUrl + '/dist/formio-sfds.standalone.js'
            console.info('injected formio-sfds:', script.src)
            document.body.appendChild(script)
            if(formioSfdsUrl) {
              formioSfdsUrl.href = script.src
            }
          })
        `
      }
    } else {
      throw new Error('No <script> found with formio-sfds in the src attribute')
    }
  }

  setFormioJSVersion (version) {
    const script = this.getScript('formiojs@')
    if (script) {
      script.setAttribute('src', script.src.replace(/formiojs@[^/]+/, `formiojs@${version}`))
    } else {
      throw new Error('No <script> found with formiojs in the src attribute')
    }
  }

  element (name, attrs = {}, children = []) {
    const el = this.document.createElement(name)
    if (attrs) {
      for (const [key, value] of Object.entries(attrs)) {
        el.setAttribute(key, value)
      }
    }
    if (typeof children === 'string') {
      el.innerHTML = children
    } else if (Array.isArray(children)) {
      for (const child of children) {
        el.appendChild(child)
      }
    }
    return el
  }

  send (res) {
    const html = this.dom.serialize()
    res.setHeader('Content-Type', 'text/html')
    res.setHeader('Content-Length', Buffer.byteLength(html))
    res.end(html, 'utf8')
  }
}
