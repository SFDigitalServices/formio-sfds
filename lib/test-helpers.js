export {
  createElement,
  createForm,
  destroyForm,
  sleep
}

function createElement (name = 'div', attrs = {}) {
  const el = document.createElement(name)
  for (const [key, value] of Object.entries(attrs)) {
    el.setAttribute(key, value)
  }
  return el
}

function createForm (form = {}, options = {}) {
  const el = createElement('div', {
    id: `form-${Date.now().toString(36)}`
  })
  document.body.appendChild(el)
  return global.Formio.createForm(el, form, options)
}

function destroyForm (form) {
  const { element } = form
  form.destroy()
  element.remove()
}

function sleep (ms = 1) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}
