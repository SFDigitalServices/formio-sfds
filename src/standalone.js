import 'sf-design-system/public/dist/css/all.css'
import '../dist/formio-sfds.css'

import 'formiojs/dist/formio.full.min.js'
import FormioSFDS from './index.js'

window.Formio.use(FormioSFDS)
