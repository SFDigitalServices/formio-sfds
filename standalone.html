<!doctype html>
<html>
  <head>
    <title>Loading...</title>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <meta name="viewport" content="width=device-width">
    <link rel="stylesheet" href="https://sf.gov/sites/default/files/css/css_tapi1dXrMXCVexRBH_J8sbMleg6570eiohsL6J_w18Y.css">
    <link rel="stylesheet" href="https://sf.gov/sites/default/files/css/css_Mq_8mlFSit3pht3tEH1HfSn5Em_cgGYwZyDnnBhhGrM.css">
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Rubik:300,400,500,700">
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noto+Sans+TC:300,400,500,700&amp;display=swap&amp;subset=chinese-traditional">
  </head>
  <body>
    <div class="page-node-type-form-page">
      <div class="form-page">
        <div id="formio" class="container p-3" data-resource="https://sfds.form.io/workers-families-first"></div>
      </div>
    </div>

    <script src="https://unpkg.com/formiojs@latest/dist/formio.full.js"></script>
    <script src="dist/formio-sfds.standalone.js"></script>
    <script>
      (() => {
        const root = document.getElementById('formio')

        const params = new URLSearchParams(location.search || location.hash.substr(1))
        const language = params.get('lang')
        const renderMode = params.get('mode')
        const hooks = tryParse(params.get('hooks'))
        const resource = params.get('res') || root.getAttribute('data-resource')
        const unlockNavigation = true

        const options = { language, renderMode, hooks, unlockNavigation, prefill: params }

        options.googleTranslate = false
        options.i18n = {
          en: { Submit: 'Join waitlist' },
          es: { Submit: '$t(Join waitlist)' },
          tl: { Submit: '$t(Join waitlist)' },
          zh: { Submit: '$t(Join waitlist)' }
        }

        Formio.createForm(root, resource, options).then(form => {
          console.log('form ready:', form)
          document.title = form.schema.title
        })

        function tryParse(str) {
          if (!str) return
          try { return JSON.parse(str) }
          catch (error) {
            console.error('Unable to parse JSON:', str)
          }
        }

      })()
    </script>
  </body>
</html>
