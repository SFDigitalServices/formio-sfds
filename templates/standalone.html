<!doctype html>
<html>
  <head>
    <title>Loading...</title>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <meta name="viewport" content="width=device-width">
    <link rel="stylesheet" href="/dist/common.css">
  </head>
  <body>
    <div class="page-node-type-form-page sfgov-translate-lang-es" lang="en">
      <div class="form-page">
        <div id="formio" class="container p-3" data-resource="https://sfds.form.io/workers-families-first"></div>
      </div>
    </div>
  </div>

  <script src="https://unpkg.com/formiojs@{{ pkg.dependencies.formiojs }}/dist/formio.full.js"></script>
  <script src="/dist/formio-sfds.standalone.js"></script>
  <script>
    (() => {
      const root = document.getElementById('formio')

      const params = new URLSearchParams(location.search || location.hash.substr(1))
      const language = params.get('lang')
      const i18n = tryParse(params.get('i18n'))
      const renderMode = params.get('mode')
      const hooks = tryParse(params.get('hooks'))
      const resource = params.get('res') || root.getAttribute('data-resource')
      const unlockNavigation = true

      const options = {
        language,
        renderMode,
        hooks,
        unlockNavigation,
        i18n,
        googleTranslate: params.get('googleTranslate') === 'true',
        prefill: params,
        properties: tryParse(params.get('properties')) || {
          backURL: "https://sf.gov/apply-building-permit",
          backTitle: "Apply for a building permit"
        }
      }

      if (language && params.get('langMode') === 'dom') {
        document.documentElement.setAttribute('lang', language)
        delete options.language
      }

      Formio.createForm(root, resource, options).then(form => {
        console.log('form ready:', form)
        document.title = form.schema.title
      })

      function tryParse(str) {
        if (!str) return
        try { return JSON.parse(str) }
        catch (error) {
          console.error('Unable to parse JSON:', str)
          return str
        }
      }

    })()
  </script>
</body>

</html>
