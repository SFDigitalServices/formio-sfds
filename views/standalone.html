<!doctype html>
<html lang="en">
  <head>
    <title>Loading...</title>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <meta name="viewport" content="width=device-width">
    <link rel="stylesheet" href="/dist/common.css">
  </head>
  <body>
    <div class="page-node-type-form-page sfgov-translate-lang-es" lang="en">
      <div class="form-page">
        <div id="formio" class="container p-3"></div>
      </div>
    </div>
  </div>

  <script src="https://unpkg.com/formiojs@{{ pkg.dependencies.formiojs }}/dist/formio.full.min.js"></script>
  <script src="/dist/formio-sfds.standalone.js"></script>
  <script>
    (() => {
      const root = document.getElementById('formio')

      const params = new URLSearchParams(location.search || location.hash.substr(1))
      const language = params.get('lang')
      const resource = params.get('res') || 'https://sfds.form.io/workers-families-first'

      if (language) {
        document.documentElement.setAttribute('lang', language)
      }

      const options = {
        language,
        renderMode: params.get('mode'),
        hooks: tryParse(params.get('hooks')),
        unlockNavigation: params.get('nav') !== 'false',
        i18n: tryParse(params.get('i18n')),
        googleTranslate: params.get('googleTranslate') === 'true',
        prefill: params
        properties: tryParse(params.get('properties'))
      }

      Formio.createForm(root, resource, options).then(form => {
        console.log('form ready:', form)
        document.title = form.schema.title
      })

      function tryParse(str) {
        if (!str) return
        try { return JSON.parse(str) }
        catch (error) {
          console.error('Unable to parse JSON:', str)
          return str
        }
      }

    })()
  </script>
</body>

</html>
