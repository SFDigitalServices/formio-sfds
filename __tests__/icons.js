/* eslint-env jest */
import { createForm, destroyForm } from '../lib/test-helpers'
import '../dist/formio-sfds.standalone.js'

describe('icons', () => {
  describe('<sfgov-icon> replacement', () => {
    it('replaces all .fa elements with <sfgov-icon>', async () => {
      const form = await createForm({
        components: [
          {
            name: 'Datetime',
            type: 'datetime',
            key: 'datetime'
          }
        ]
      })

      const faElements = form.element.querySelectorAll('.fa')
      expect(faElements).toHaveLength(0)
      const faCalendar = form.element.querySelectorAll('.fa-calendar')
      expect(faCalendar).toHaveLength(0)

      const icons = form.element.querySelectorAll('sfgov-icon')
      expect(icons).toHaveLength(1)
      expect(icons[0].getAttribute('symbol')).toBe('calendar')
    })

    it('preserves legacy "square" icons', async () => {
      const form = await createForm({
        components: [
          {
            name: 'Text',
            type: 'htmlelement',
            key: 'text',
            content: '<span data-icon="square"></span>'
          }
        ]
      })
      expect(form.element.querySelector('[ref=html]').innerHTML).toMatchSnapshot()
    })

    it('removes .fa and .fa-{symbol} classes', async () => {
      const form = await createForm({
        components: [
          {
            name: 'Text',
            type: 'htmlelement',
            key: 'text',
            content: '<i class="fa fa-camera"></i>'
          }
        ]
      })
      expect(form.element.querySelectorAll('.fa, .fa-camera')).toHaveLength(0)
    })
  })

  describe('datetime prefix', () => {
    it('renders a calendar icon', async () => {
      const form = await createForm({
        type: 'form',
        display: 'form',
        components: [
          {
            type: 'datetime',
            label: 'Date'
          }
        ]
      }, {})

      const prepend = form.element.querySelector('.input-group-prepend')
      expect(prepend).not.toBe(null)
      expect(prepend.innerHTML).not.toContain('[object HTMLElement]')

      destroyForm(form)
    })
  })
})
