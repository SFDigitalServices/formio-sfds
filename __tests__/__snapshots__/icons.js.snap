// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icons <sfgov-icon> replacement preserves legacy "square" icons 1`] = `
<span data-icon="square"
      class="notranslate"
>
  <svg viewbox="0 0 20 20"
       fill="currentColor"
       height="20"
  >
    <rect width="20"
          height="20"
          rx="4"
    >
    </rect>
  </svg>
</span>
`;
